packages/repository/src/repository-builder.ts

```diff
@@ -0,0 +1,163 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/rest-crud
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import assert from 'assert';
+import {Constructor} from './common-types';
+import {Entity, Model} from './model';
+import {
+  CrudRepository,
+  DefaultCrudRepository,
+  DefaultKeyValueRepository,
+  EntityCrudRepository,
+  juggler,
+  KeyValueRepository,
+} from './repositories';
+
+/**
+ * Create (define) an entity CRUD repository class for the given model.
+ *
+ * @example
+ *
+ * ```ts
+ * const ProductRepository = defineEntityCrudRepositoryClass(Product);
+ * ```
+ *
+ * @param entityClass - An entity class such as `Product`.
+ * @param baseRepositoryClass - Base repository class. Defaults to `DefaultCrudRepository`
+ *
+ * @typeParam E - An entity class
+ * @typeParam IdType - ID type for the entity
+ * @typeParam Relations - Relations for the entity
+ */
+export function defineEntityCrudRepositoryClass<
+  T extends Entity,
+  IdType,
+  Relations extends object = {}
+>(
+  entityClass: typeof Entity & {prototype: T},
+  baseRepositoryClass: Constructor<
+    EntityCrudRepository<T, IdType, Relations>
+  > = DefaultCrudRepository as Constructor<
+    EntityCrudRepository<T, IdType, Relations>
+  >,
+): EntityCrudRepositoryClass<T, IdType, Relations> {
+  const repoName = entityClass.name + 'Repository';
+  const defineNamedRepo = new Function(
+    'EntityCtor',
+    'BaseRepository',
+    `return class ${repoName} extends BaseRepository {
+      constructor(dataSource) {
+        super(EntityCtor, dataSource);
+      }
+    };`,
+  );
+
+  const repo = defineNamedRepo(entityClass, baseRepositoryClass);
+  assert.equal(repo.name, repoName);
+  return repo;
+}
+
+/**
+ * Signature for EntityCrudRepository classes
+ *
+ * @typeParam E - An entity class
+ * @typeParam IdType - ID type for the entity
+ * @typeParam Relations - Relations for the entity
+ */
+export interface EntityCrudRepositoryClass<
+  E extends Entity,
+  IdType,
+  Relations extends object
+> {
+  new (ds: juggler.DataSource): EntityCrudRepository<E, IdType, Relations>;
+}
+
+/**
+ * Create (define) a CRUD repository class for the given model.
+ *
+ * @example
+ *
+ * ```ts
+ * const AddressRepository = defineCrudRepositoryClass(Address);
+ * ```
+ *
+ * @param modelClass - A model class such as `Address`.
+ *
+ * @typeParam M - Model class
+ */
+export function defineCrudRepositoryClass<M extends Model>(
+  modelClass: typeof Model & {prototype: M},
+  baseRepositoryClass: Constructor<CrudRepository<M>>,
+): CrudRepositoryClass<M> {
```

Comment:
IMO, we need to make the interface of the repository class configurable. Ideally, `defineEntityCrudRepositoryClass` should delegate most of the implementation to `defineCrudRepositoryClass`.

A mock-up, I am not sure if it will work:

```ts

export function defineCrudRepositoryClass<
  M extends Model, 
  R extends Constructor<CrudRepository<M>>
>(
   modelClass: typeof Model & {prototype: M},
   baseRepositoryClass: R,
 ): R {
  // ...
}

export function defineEntityCrudRepositoryClass<
   T extends Entity,
   IdType,
   Relations extends object = {},
   R extends Constructor<EntityCrudRepository<T, IdType, Relations>>
    = Constructor<EntityCrudRepository<T, IdType, Relations>
 >(
   entityClass: typeof Entity & {prototype: T},
   baseRepositoryClass: R = DefaultCrudRepository as Constructor<
     EntityCrudRepository<T, IdType, Relations>
   >,
 ): R {
  return defineCrudRepositoryClass<M, R>(entityClass, baseRepositoryClass);
}
```

A test case to verify that extra API added by custom base classes are correctly propagated at compile time:

```ts
class TitledEntityCrudRepository<
   T extends Entity & {title: string},
   IdType,
   Relations extends object = {},
> extends DefaultCrudRepository<T, IdType, Relations> {
  findByTitle(title: string) {
    return this.find({where: {title}});
  }
}

@model()
class Note extends Entity {
  @property({id: true})
  id: number;

  @property()
  title: string;
}

const NoteRepository = defineEntityCrudRepositoryClass(Note, TitledEntityCrudRepository);
const db = new juggler.DataSource({connector: 'memory'});
const repo = new NoteRepository(db);

const created = await repo.create({title: 'welcome'});
const found = await repo.findByTitle('welcome');
expect(found).to.deepEqual(created);
```